File: desktop_app/src/inferno_analyzer/driver/analysis/driver_consistency.py
"""Analysis pipeline for driver consistency across corners.

Orchestrates existing motorsports_data_notebook functions into a
corner-by-corner consistency analysis.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import TYPE_CHECKING

import numpy as np
import pandas as pd

from motorsports_data_notebook.corners import Corner, identify_corners
from motorsports_data_notebook.driver_analysis import (
    find_throttle_acceptance_from_arrays,
    prepare_throttle_acceptance,
)
from motorsports_data_notebook.zones import (
    TrackSegment,
    compute_segment_stats_from_arrays,
    create_track_segments,
    detect_zones_from_arrays,
)

if TYPE_CHECKING:
    from motorsports_data_notebook._types import LogFile


@dataclass
class LapTraceData:
    """Per-lap trace data for detail overlay visualization.

    Attributes
    ----------
    lap_num : int
        Lap number.
    distance : np.ndarray
        Distance values in meters.
    throttle : np.ndarray
        Throttle position values (%).
    brake : np.ndarray
        Brake pressure values.
    lateral_g : np.ndarray
        Lateral G values.
    """

    lap_num: int
    distance: np.ndarray
    throttle: np.ndarray
    brake: np.ndarray
    lateral_g: np.ndarray


@dataclass
class CornerConsistencyData:
    """Consistency data for a single corner across multiple laps.

    Attributes
    ----------
    corner : Corner
        The corner definition.
    ta_values : list[float]
        Throttle acceptance % per lap (excluding None results).
    ta_mean : float
        Mean throttle acceptance %.
    ta_std : float
        Std dev of throttle acceptance %.
    bp_values : list[float]
        Braking distance per lap.
    bp_std : float
        Braking point std dev (lower = more consistent).
    speed_values : list[float]
        Minimum corner speed per lap.
    speed_mean : float
        Mean minimum corner speed.
    speed_std : float
        Std dev of minimum corner speed.
    lap_traces : list[LapTraceData]
        Per-lap trace data for detail overlay.
    braking_start : float | None
        Start distance of braking zone (for detail view markers).
    """

    corner: Corner
    ta_values: list[float] = field(default_factory=list)
    ta_mean: float = 0.0
    ta_std: float = 0.0
    bp_values: list[float] = field(default_factory=list)
    bp_std: float = 0.0
    speed_values: list[float] = field(default_factory=list)
    speed_mean: float = 0.0
    speed_std: float = 0.0
    exit_speed_values: list[float] = field(default_factory=list)
    exit_speed_mean: float = 0.0
    exit_speed_std: float = 0.0
    accel_zone_length: float = 0.0
    opportunity_score: float = 0.0
    lap_traces: list[LapTraceData] = field(default_factory=list)
    braking_start: float | None = None


@dataclass
class DriverConsistencyResult:
    """Complete result of driver consistency analysis.

    Attributes
    ----------
    corners : list[Corner]
        Detected corners.
    corner_data : list[CornerConsistencyData]
        Per-corner consistency data.
    segments : list[TrackSegment]
        Track segments.
    stats_df : pd.DataFrame
        Per-lap segment statistics.
    ref_lat : np.ndarray
        Reference lap GPS latitude.
    ref_lon : np.ndarray
        Reference lap GPS longitude.
    ref_distance : np.ndarray
        Reference lap distance in meters.
    """

    corners: list[Corner]
    corner_data: list[CornerConsistencyData]
    segments: list[TrackSegment]
    stats_df: pd.DataFrame
    ref_lat: np.ndarray
    ref_lon: np.ndarray
    ref_distance: np.ndarray


def analyze_driver_consistency(
    log: "LogFile",
    selected_laps: list[int],
    channel_names: dict[str, str],
    corner_threshold: float = 0.006,
    throttle_threshold: float = 98.0,
    sustain_time_ms: float = 500.0,
) -> DriverConsistencyResult:
    """Run the full driver consistency analysis pipeline.

    Steps:
    1. Get best lap GPS data for corner detection
    2. Detect corners from GPS curvature
    3. Detect braking/acceleration zones across selected laps
    4. Create track segments combining corners and zones
    5. Compute per-lap segment statistics
    6. Compute throttle acceptance per lap per corner
    7. Extract per-lap trace data for detail overlays
    8. Aggregate into CornerConsistencyData

    Parameters
    ----------
    log : LogFile
        Loaded and enriched session data.
    selected_laps : list[int]
        Lap numbers to include in analysis.
    channel_names : dict[str, str]
        Channel name mapping with keys: throttle, brake, lateral_g,
        gps_lat, gps_lon, gps_speed.
    corner_threshold : float, default=0.006
        Curvature threshold for corner detection.
    throttle_threshold : float, default=98.0
        Throttle % threshold for throttle acceptance.
    sustain_time_ms : float, default=500.0
        Time in ms that throttle must be sustained.

    Returns
    -------
    DriverConsistencyResult
        Complete analysis results.
    """
    laps_df = log.laps.to_pandas()

    # Filter to selected laps
    selected_laps_df = laps_df[laps_df["num"].isin(selected_laps)].copy()
    if len(selected_laps_df) == 0:
        raise ValueError("No valid laps selected")

    # Step 1: Get best lap GPS for corner detection
    gps_channels = [channel_names["gps_lat"], channel_names["gps_lon"], "distance_m"]
    # Use fastest selected lap for GPS reference (avoids out-laps / incomplete laps)
    selected_laps_df["_duration"] = selected_laps_df["end_time"] - selected_laps_df["start_time"]
    best_lap = selected_laps_df.loc[selected_laps_df["_duration"].idxmin()]
    best_lap_num = int(best_lap["num"])
    best_lap_log = log.filter_by_lap(best_lap_num)
    best_channels = (
        best_lap_log.select_channels(gps_channels)
        .resample_to_channel(channel_names["gps_lat"])
        .channels
    )
    lat_table = best_channels[channel_names["gps_lat"]]
    lon_table = best_channels[channel_names["gps_lon"]]
    lat = lat_table.column(channel_names["gps_lat"]).to_numpy()
    lon = lon_table.column(channel_names["gps_lon"]).to_numpy()
    dist_table = best_channels["distance_m"]
    ref_dist = dist_table.column("distance_m").to_numpy()

    # Filter ref_dist to match valid GPS samples (identify_corners filters lat/lon internally)
    valid_gps = (lat != 0.0) | (lon != 0.0)
    if not np.all(valid_gps):
        ref_dist = ref_dist[valid_gps]

    # Step 2: Detect corners
    corners = identify_corners(lat, lon, threshold=corner_threshold)
    if not corners:
        raise ValueError("No corners detected. Try adjusting the corner detection threshold.")

    # Get track length from best lap distance
    track_length = float(ref_dist[-1])

    # Step 3-5: Single-pass per-lap extraction + zone detection + segment stats + TA
    throttle_ch = channel_names["throttle"]
    brake_ch = channel_names["brake"]
    lateral_g_ch = channel_names["lateral_g"]
    gps_speed_ch = channel_names["gps_speed"]

    # Build deduplicated channel list for extraction
    all_channels = ["distance_m", throttle_ch, brake_ch, lateral_g_ch, "speed_kmh", gps_speed_ch]
    all_channels = list(dict.fromkeys(all_channels))

    # Per-lap caches
    lap_distances: list[np.ndarray] = []
    lap_speeds: list[np.ndarray] = []
    lap_brakes: list[np.ndarray] = []
    lap_throttles: list[np.ndarray] = []
    lap_gps_speeds: list[np.ndarray] = []
    lap_lateral_gs: list[np.ndarray] = []
    lap_timecodes_list: list[np.ndarray] = []
    lap_num_list: list[int] = []
    lap_time_list: list[float] = []

    for _, lap_row in selected_laps_df.iterrows():
        lap_num = int(lap_row["num"])
        try:
            aligned = (
                log.filter_by_lap(lap_num)
                .select_channels(all_channels)
                .resample_to_channel("distance_m")
                .channels
            )
        except Exception:
            continue

        ref_table = aligned["distance_m"]
        distance = ref_table.column("distance_m").to_numpy()
        if len(distance) == 0:
            continue

        lap_distances.append(distance)
        lap_timecodes_list.append(ref_table.column("timecodes").to_numpy())
        lap_brakes.append(aligned[brake_ch].column(brake_ch).to_numpy())
        lap_throttles.append(aligned[throttle_ch].column(throttle_ch).to_numpy())
        lap_speeds.append(aligned["speed_kmh"].column("speed_kmh").to_numpy())
        lap_gps_speeds.append(aligned[gps_speed_ch].column(gps_speed_ch).to_numpy())
        lap_lateral_gs.append(aligned[lateral_g_ch].column(lateral_g_ch).to_numpy())
        lap_num_list.append(lap_num)
        lap_time_list.append(lap_row["lap_time"])

    # Zone detection (no filter_by_lap)
    braking_zones, accel_zones = detect_zones_from_arrays(
        lap_distances, lap_brakes, lap_throttles, lap_gps_speeds,
    )

    # Create segments (unchanged)
    segments = create_track_segments(corners, braking_zones, accel_zones, track_length)

    # Segment stats (no filter_by_lap)
    stats_df = compute_segment_stats_from_arrays(
        lap_distances, lap_speeds, lap_brakes, lap_throttles,
        lap_num_list, lap_time_list, segments,
    )

    # Pre-compute per-corner metadata
    corner_data_list: list[CornerConsistencyData] = []
    trace_bounds: list[tuple[float, float]] = []

    for corner in corners:
        cd = CornerConsistencyData(corner=corner)

        corner_braking = stats_df[
            (stats_df["corner_id"] == corner.id) & (stats_df["segment_type"] == "braking")
        ]
        if len(corner_braking) > 0 and "braking_point" in corner_braking.columns:
            bp_vals = corner_braking["braking_point"].dropna().tolist()
            if bp_vals:
                cd.bp_values = bp_vals
                cd.bp_std = float(np.std(bp_vals))

        corner_seg = stats_df[
            (stats_df["corner_id"] == corner.id) & (stats_df["segment_type"] == "corner")
        ]
        if len(corner_seg) > 0 and "min_speed" in corner_seg.columns:
            speed_vals = corner_seg["min_speed"].dropna().tolist()
            if speed_vals:
                cd.speed_values = speed_vals
                cd.speed_mean = float(np.mean(speed_vals))
                cd.speed_std = float(np.std(speed_vals))

        if len(corner_seg) > 0 and "exit_speed" in corner_seg.columns:
            exit_vals = corner_seg["exit_speed"].dropna().tolist()
            if exit_vals:
                cd.exit_speed_values = exit_vals
                cd.exit_speed_mean = float(np.mean(exit_vals))
                cd.exit_speed_std = float(np.std(exit_vals))

        accel_seg = next(
            (s for s in segments if s.corner_id == corner.id and s.segment_type == "acceleration"),
            None,
        )
        if accel_seg is not None:
            cd.accel_zone_length = accel_seg.length
            if cd.exit_speed_std > 0:
                cd.opportunity_score = cd.exit_speed_std * cd.accel_zone_length

        braking_seg = next(
            (s for s in segments if s.corner_id == corner.id and s.segment_type == "braking"),
            None,
        )
        trace_margin_before = 50.0
        if braking_seg is not None:
            trace_start = braking_seg.start_dist - trace_margin_before
            cd.braking_start = braking_seg.start_dist
        else:
            trace_start = corner.start_dist - 150.0
        trace_end = corner.end_dist + 50.0
        trace_bounds.append((trace_start, trace_end))

        corner_data_list.append(cd)

    # Per-lap TA + trace extraction (using cached arrays, no filter_by_lap)
    for i, (lap_distance, lap_timecodes, lap_throttle, lap_brake, lap_lateral_g) in enumerate(
        zip(lap_distances, lap_timecodes_list, lap_throttles, lap_brakes, lap_lateral_gs)
    ):
        lap_num = lap_num_list[i]

        # Prepare TA shared data once per lap
        smoothed, eff_thresh = prepare_throttle_acceptance(
            lap_throttle, lap_lateral_g, throttle_threshold=throttle_threshold,
        )

        for ci, corner in enumerate(corners):
            trace_start, trace_end = trace_bounds[ci]
            si = np.searchsorted(lap_distance, trace_start)
            ei = np.searchsorted(lap_distance, trace_end, side="right")
            if si >= ei:
                continue

            corner_data_list[ci].lap_traces.append(
                LapTraceData(
                    lap_num=lap_num,
                    distance=lap_distance[si:ei].copy(),
                    throttle=lap_throttle[si:ei].copy(),
                    brake=lap_brake[si:ei].copy(),
                    lateral_g=lap_lateral_g[si:ei].copy(),
                )
            )

            result = find_throttle_acceptance_from_arrays(
                lap_distance, lap_timecodes, lap_throttle, lap_lateral_g, corner,
                smoothed_lateral_g=smoothed, effective_threshold=eff_thresh,
                sustain_time_ms=sustain_time_ms,
            )
            if result is not None:
                corner_data_list[ci].ta_values.append(result["throttle_acceptance_pct"])

    # Post-loop: compute TA aggregates
    for cd in corner_data_list:
        if cd.ta_values:
            cd.ta_mean = float(np.mean(cd.ta_values))
            cd.ta_std = float(np.std(cd.ta_values))

    return DriverConsistencyResult(
        corners=corners,
        corner_data=corner_data_list,
        segments=segments,
        stats_df=stats_df,
        ref_lat=lat,
        ref_lon=lon,
        ref_distance=ref_dist,
    )

File: desktop_app_rs/crates/inferno-core/src/analysis/driver_consistency.rs
use rayon::prelude::*;

use super::corners::{self, Corner};
use super::math;
use super::throttle_acceptance;
use super::zones::{self, SegmentType, TrackSegment};
use crate::channel;
use crate::error::{Error, Result};
use crate::lap::Lap;
use crate::session::{LapData, Session};

/// Per-lap trace data for visualization overlays.
#[derive(Debug, Clone)]
pub struct LapTraceData {
    pub lap_num: i32,
    pub distance: Vec<f64>,
    pub throttle: Vec<f64>,
    pub brake: Vec<f64>,
    pub lateral_g: Vec<f64>,
}

/// Per-corner aggregated consistency data.
#[derive(Debug, Clone)]
pub struct CornerConsistencyData {
    pub corner: Corner,
    pub ta_values: Vec<f64>,
    pub ta_mean: f64,
    pub ta_std: f64,
    pub bp_values: Vec<f64>,
    pub bp_std: f64,
    pub speed_values: Vec<f64>,
    pub speed_mean: f64,
    pub speed_std: f64,
    pub exit_speed_values: Vec<f64>,
    pub exit_speed_mean: f64,
    pub exit_speed_std: f64,
    pub accel_zone_length: f64,
    pub opportunity_score: f64,
    pub lap_traces: Vec<LapTraceData>,
    pub braking_start: Option<f64>,
}

/// Full driver consistency analysis result.
pub struct DriverConsistencyResult {
    pub corners: Vec<Corner>,
    pub corner_data: Vec<CornerConsistencyData>,
    pub segments: Vec<TrackSegment>,
    pub ref_lat: Vec<f64>,
    pub ref_lon: Vec<f64>,
    pub ref_distance: Vec<f64>,
}

/// Channel name mapping for analysis.
pub struct ChannelConfig {
    pub throttle: String,
    pub brake: String,
    pub lateral_g: String,
    pub gps_lat: String,
    pub gps_lon: String,
    pub gps_speed: String,
}

impl Default for ChannelConfig {
    fn default() -> Self {
        Self {
            throttle: "PPS".into(),
            brake: "BrakePress".into(),
            lateral_g: "LateralAcc".into(),
            gps_lat: "GPS Latitude".into(),
            gps_lon: "GPS Longitude".into(),
            gps_speed: "GPS Speed".into(),
        }
    }
}

/// Run the full driver consistency analysis pipeline.
pub fn analyze_driver_consistency(
    session: &Session,
    selected_laps: &[i32],
    config: &ChannelConfig,
    corner_threshold: f64,
    throttle_threshold: f64,
    sustain_time_ms: f64,
) -> Result<DriverConsistencyResult> {
    // Find selected laps
    let laps: Vec<&Lap> = session
        .laps
        .iter()
        .filter(|l| selected_laps.contains(&l.num))
        .collect();

    if laps.is_empty() {
        return Err(Error::NoValidLaps);
    }

    // Get best lap for GPS reference
    let best_lap = laps
        .iter()
        .filter(|l| l.duration_ms() > 0)
        .min_by_key(|l| l.duration_ms())
        .ok_or(Error::NoValidLaps)?;

    // Extract GPS data from best lap
    let channel_names = [
        config.gps_lat.as_str(),
        config.gps_lon.as_str(),
        config.gps_speed.as_str(),
        config.throttle.as_str(),
        config.brake.as_str(),
        config.lateral_g.as_str(),
        "speed_kmh",
    ];

    let ref_data = LapData::extract(session, best_lap, &channel_names)?;
    let ref_lat = ref_data
        .get(&config.gps_lat)
        .ok_or_else(|| Error::MissingChannel(config.gps_lat.clone()))?;
    let ref_lon = ref_data
        .get(&config.gps_lon)
        .ok_or_else(|| Error::MissingChannel(config.gps_lon.clone()))?;

    // Filter valid GPS
    let valid_mask: Vec<bool> = ref_lat
        .iter()
        .zip(ref_lon.iter())
        .map(|(&lat, &lon)| lat != 0.0 || lon != 0.0)
        .collect();

    let valid_lat: Vec<f64> = ref_lat
        .iter()
        .zip(&valid_mask)
        .filter(|(_, &m)| m)
        .map(|(&v, _)| v)
        .collect();
    let valid_lon: Vec<f64> = ref_lon
        .iter()
        .zip(&valid_mask)
        .filter(|(_, &m)| m)
        .map(|(&v, _)| v)
        .collect();

    // Identify corners
    let corners = corners::identify_corners(&valid_lat, &valid_lon, corner_threshold);
    if corners.is_empty() {
        return Err(Error::Other("No corners detected".into()));
    }

    // Extract per-lap data (parallelized)
    let lap_data: Vec<LapData> = laps
        .par_iter()
        .filter_map(|lap| LapData::extract(session, lap, &channel_names).ok())
        .collect();

    if lap_data.is_empty() {
        return Err(Error::NoValidLaps);
    }

    // Check which channels are available (brake/throttle are optional for GPS-only loggers)
    if !lap_data.iter().any(|ld| ld.get("speed_kmh").is_some()) {
        let available: Vec<&str> = session.channel_names();
        return Err(Error::MissingChannel(format!(
            "speed_kmh (derived) not found. Available: {}",
            available.join(", ")
        )));
    }

    // Collect arrays for zone detection
    let distances: Vec<Vec<f64>> = lap_data.iter().map(|ld| ld.dist().to_vec()).collect();
    let brake_presses: Vec<Vec<f64>> = lap_data
        .iter()
        .map(|ld| ld.get(&config.brake).unwrap_or(&[]).to_vec())
        .collect();
    let throttles: Vec<Vec<f64>> = lap_data
        .iter()
        .map(|ld| ld.get(&config.throttle).unwrap_or(&[]).to_vec())
        .collect();
    let speeds: Vec<Vec<f64>> = lap_data
        .iter()
        .map(|ld| ld.get("speed_kmh").unwrap_or(&[]).to_vec())
        .collect();

    // Detect zones
    let (braking_zones, accel_zones) = zones::detect_zones_from_arrays(
        &distances,
        &brake_presses,
        &throttles,
        &speeds,
        1.0,
        0.5,
        1.5,
        None,
        None,
    );

    // Create segments
    let track_length = distances
        .iter()
        .filter_map(|d| d.last().copied())
        .fold(0.0f64, f64::max);

    let segments =
        zones::create_track_segments(&corners, &braking_zones, &accel_zones, track_length);

    // Compute segment stats
    let dist_refs: Vec<&[f64]> = distances.iter().map(|d| d.as_slice()).collect();
    let speed_refs: Vec<&[f64]> = speeds.iter().map(|s| s.as_slice()).collect();
    let brake_refs: Vec<&[f64]> = brake_presses.iter().map(|b| b.as_slice()).collect();
    let lap_nums: Vec<i32> = lap_data.iter().map(|ld| ld.lap_num).collect();

    // Auto-detect brake threshold for stats
    let brake_threshold = brake_presses
        .first()
        .map(|b| {
            let max_brake = b.iter().cloned().fold(f64::NEG_INFINITY, f64::max);
            0.05 * max_brake
        })
        .unwrap_or(5.0);

    let segment_stats = zones::compute_segment_stats_from_arrays(
        &dist_refs,
        &speed_refs,
        &brake_refs,
        &lap_nums,
        &segments,
        brake_threshold,
    );

    // Build per-corner consistency data
    let trace_margin = 50.0;
    let mut corner_data = Vec::with_capacity(corners.len());

    for corner in &corners {
        // Collect braking point stats
        let bp_values: Vec<f64> = segment_stats
            .iter()
            .filter(|s| s.corner_id == corner.id && s.segment_type == SegmentType::Braking)
            .filter_map(|s| s.braking_point)
            .collect();
        let bp_std = math::std_dev(&bp_values);

        // Collect speed stats
        let speed_values: Vec<f64> = segment_stats
            .iter()
            .filter(|s| s.corner_id == corner.id && s.segment_type == SegmentType::Corner)
            .filter_map(|s| s.min_speed)
            .collect();
        let speed_mean = math::mean(&speed_values);
        let speed_std = math::std_dev(&speed_values);

        // Collect exit speed stats
        let exit_speed_values: Vec<f64> = segment_stats
            .iter()
            .filter(|s| s.corner_id == corner.id && s.segment_type == SegmentType::Corner)
            .filter_map(|s| s.exit_speed)
            .collect();
        let exit_speed_mean = math::mean(&exit_speed_values);
        let exit_speed_std = math::std_dev(&exit_speed_values);

        // Acceleration zone length
        let accel_zone_length = segments
            .iter()
            .find(|s| s.corner_id == corner.id && s.segment_type == SegmentType::Acceleration)
            .map(|s| s.length())
            .unwrap_or(0.0);

        let opportunity_score = exit_speed_std * accel_zone_length;

        // Braking start distance
        let braking_start = segments
            .iter()
            .find(|s| s.corner_id == corner.id && s.segment_type == SegmentType::Braking)
            .map(|s| s.start_dist);

        // Extract lap traces for this corner
        let trace_start = braking_start
            .unwrap_or(corner.start_dist)
            .min(corner.start_dist)
            - trace_margin;
        let trace_end = corner.end_dist + accel_zone_length + trace_margin;

        let mut lap_traces = Vec::new();
        let mut ta_values = Vec::new();

        for ld in &lap_data {
            let dist = ld.dist();
            let si = math::searchsorted(dist, trace_start);
            let ei = math::searchsorted_right(dist, trace_end).min(dist.len());

            if si >= ei {
                continue;
            }

            lap_traces.push(LapTraceData {
                lap_num: ld.lap_num,
                distance: dist[si..ei].to_vec(),
                throttle: ld
                    .get(&config.throttle)
                    .map(|a| a[si..ei].to_vec())
                    .unwrap_or_default(),
                brake: ld
                    .get(&config.brake)
                    .map(|a| a[si..ei].to_vec())
                    .unwrap_or_default(),
                lateral_g: ld
                    .get(&config.lateral_g)
                    .map(|a| a[si..ei].to_vec())
                    .unwrap_or_default(),
            });

            // Compute throttle acceptance for this corner+lap
            if let (Some(throttle_arr), Some(lat_g_arr)) =
                (ld.get(&config.throttle), ld.get(&config.lateral_g))
            {
                let timecodes_batch = session.channels.get("distance_m").and_then(|b| {
                    channel::filter_by_lap(
                        b,
                        session.laps.iter().find(|l| l.num == ld.lap_num).unwrap(),
                    )
                    .ok()
                });

                if let Some(tc_batch) = timecodes_batch {
                    if let Ok(tc) = channel::get_timecodes(&tc_batch) {
                        let tc_slice: Vec<i64> = tc.values().to_vec();

                        let (smoothed, eff_threshold) =
                            throttle_acceptance::prepare_throttle_acceptance(
                                throttle_arr,
                                lat_g_arr,
                                Some(throttle_threshold),
                                25,
                            );

                        if let Some(ta) = throttle_acceptance::find_throttle_acceptance(
                            dist,
                            &tc_slice,
                            throttle_arr,
                            &smoothed,
                            eff_threshold,
                            corner,
                            sustain_time_ms,
                        ) {
                            ta_values.push(ta.throttle_acceptance_pct);
                        }
                    }
                }
            }
        }

        let ta_mean = math::mean(&ta_values);
        let ta_std = math::std_dev(&ta_values);

        corner_data.push(CornerConsistencyData {
            corner: corner.clone(),
            ta_values,
            ta_mean,
            ta_std,
            bp_values,
            bp_std,
            speed_values,
            speed_mean,
            speed_std,
            exit_speed_values,
            exit_speed_mean,
            exit_speed_std,
            accel_zone_length,
            opportunity_score,
            lap_traces,
            braking_start,
        });
    }

    Ok(DriverConsistencyResult {
        corners,
        corner_data,
        segments,
        ref_lat: valid_lat,
        ref_lon: valid_lon,
        ref_distance: ref_data.dist().to_vec(),
    })
}
